sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UseUriInsteadOfString.cs

```diff
@@ -51,75 +53,143 @@ public sealed class UseUriInsteadOfString : SonarDiagnosticAnalyzer
+            }
+
+            var methodOverloads = FindOverloadsThatUseUriTypeInPlaceOfString(methodSymbol, stringUrlParams);
+            if (methodOverloads.Any())
+            {
+                var methodOverloadSet = new HashSet<IMethodSymbol>(methodOverloads);
+                if (!methodDeclaration.IsKind(SyntaxKind.ConstructorDeclaration) &&
+                    !SyntaxHelper.ContainsMethodInvocation(methodDeclaration, context.SemanticModel,
+                        mSyntax => true,
+                        mSymbol => methodOverloadSet.Contains(mSymbol)))
                 {
-                    var propertyDeclaration = (PropertyDeclarationSyntax)c.Node;
-                    var propertySymbol = c.SemanticModel.GetDeclaredSymbol(propertyDeclaration);
+                    context.ReportDiagnostic(Diagnostic.Create(rule_S3997, FindIdentifierLocation(methodDeclaration)));
+                }
+            }
+            else
+            {
+                foreach (int paramIdx in stringUrlParams)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(rule_S3994,
+                        methodDeclaration.ParameterList.Parameters[paramIdx].Type.GetLocation()));
+                }
+            }
+        }
 
-                    if (propertySymbol.Type.Is(KnownType.System_String) &&
-                        !propertySymbol.IsOverride &&
-                        NameContainsUri(propertySymbol.Name))
-                    {
-                        c.ReportDiagnostic(Diagnostic.Create(rule_S3996, propertyDeclaration.Type.GetLocation()));
-                    }
-                },
-                SyntaxKind.PropertyDeclaration);
+        private Location FindIdentifierLocation(BaseMethodDeclarationSyntax baseMethodDeclaration)
+        {
+            var location = (baseMethodDeclaration as MethodDeclarationSyntax)?.Identifier.GetLocation();
+            if (location == null)
+            {
+                location = (baseMethodDeclaration as ConstructorDeclarationSyntax)?.Identifier.GetLocation();
+            }
+
+            return location;
         }
 
-        private bool HasOverloadThatUsesUriTypeInPlaceOfString(IMethodSymbol originalMethodSymbol, ISet<int> paramIdx)
+        private void VerifyReturnType(SyntaxNodeAnalysisContext context,
+            BaseMethodDeclarationSyntax methodDeclaration, IMethodSymbol methodSymbol)
         {
+            var returnTypeLocation = (methodDeclaration as MethodDeclarationSyntax)?.ReturnType?.GetLocation();
+
+            if (returnTypeLocation != null &&
+                methodSymbol.ReturnType.Is(KnownType.System_String) &&
+                NameContainsUri(methodSymbol.Name))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(rule_S3995, returnTypeLocation));
+            }
+        }
+
+        private IEnumerable<IMethodSymbol> FindOverloadsThatUseUriTypeInPlaceOfString(IMethodSymbol originalMethodSymbol, ISet<int> paramIdx)
```

Comment:
I think we should be a bit more permissive and allow calling any overload that has Uri argument to suppress our issue, regardless of the argument positions. In methods with many arguments it is good but not required the arguments to stay at the same index and we could generate noise.